### ISSUE TITLE
Is caption selection broken?

### ISSUE THREAD

**Issue by mathiasi** [googlecast/CastCompanionLibrary-android#330]
I feel like I have tried everything now so I need to reach out for some help.
I have tried to get captions to work but with no luck so far. I can get it to display the CC-selection icon and it displays my registered caption but when I select it and press OK, nothing shows up and when I open up the selection dialog again, the selection is back on "None".
This is the console output from once I hit OK in the dialog:
```
08-10 14:19:27.558 3579-3579/com.example.mathias.cast_subtitles_test I/ViewRootImpl: ViewRoot's Touch Event : ACTION_DOWN
08-10 14:19:27.602 3579-3579/com.example.mathias.cast_subtitles_test I/ViewRootImpl: ViewRoot's Touch Event : ACTION_UP
08-10 14:19:27.632 3579-3719/com.example.mathias.cast_subtitles_test E/Surface: getSlotFromBufferLocked: unknown buffer: 0x964ed410
08-10 14:19:27.646 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] Setting track result was successful? true
08-10 14:19:27.663 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] RemoteMediaPlayer::onStatusUpdated() is reached
08-10 14:19:27.663 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] onRemoteMediaPlayerStatusUpdated() reached
08-10 14:19:27.663 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] onQueueUpdated() reached
08-10 14:19:27.663 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] Queue Items size: 1, Item: com.google.android.gms.cast.MediaQueueItem@d961ab5, Repeat Mode: 0, Shuffle: false
08-10 14:19:27.664 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] onRemoteMediaPlayerStatusUpdated(): Player status = paused
08-10 14:19:27.668 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] startNotificationService()
08-10 14:19:27.670 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] updateMiniControllersVisibility() reached with visibility: true
08-10 14:19:27.671 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastControlle: [v2.8.2] updatePlayerStatus(), state: 3
08-10 14:19:27.672 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] RemoteMediaPlayer::onMetadataUpdated() is reached
08-10 14:19:27.672 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastManager: [v2.8.2] onRemoteMediaPlayerMetadataUpdated() reached
08-10 14:19:27.786 3579-3590/com.example.mathias.cast_subtitles_test I/art: Background partial concurrent mark sweep GC freed 19894(830KB) AllocSpace objects, 11(8MB) LOS objects, 40% free, 37MB/62MB, paused 414us total 112.825ms
08-10 14:19:27.870 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastNotificat: [v2.8.2] onStartCommand
08-10 14:19:27.870 3579-3579/com.example.mathias.cast_subtitles_test D/ccl_VideoCastNotificat: [v2.8.2] onStartCommand(): Action: ACTION_VISIBILITY false
```

The "Setting track result was successful? true" seems to me that there is something going on but nothing shows up on the screen and the option reverts back.
I'm using the default receiver application but I have also tried with a custom (just a clone of the default) so that I could see debug messages, but nothing shows up in the log.
I know that lots of people run into issues with CORS, but I have checked that my captions file (a VTT file) complies with CORS. I have manually edited the player.html and added the VTT file as a track and it shows up just fine.

I'm working on a blank app so most of the code is contained the Main activity which can be seen here: https://gist.github.com/mathiasi/620cf4c60c93201910abde46cb740fb5 but the full project can be found here:
[Cast-Subtitles-Test.zip](https://github.com/googlecast/CastCompanionLibrary-android/files/411036/Cast-Subtitles-Test.zip)

I hope someone has a suggestion to what could be wrong because right now I'm at a state of despair of how to proceed.

-Thanks!

**Comment by naddaf** [comment]
Make sure that your mp4 media is also on a server that supports CORS; when you use tracks/closed captions, your track **and** your media (even mp4) need to both have CORS headers.

**Comment by mathiasi** [comment]
I have just checked and both resources: the mp4 and vtt responds with header "Access-Control-Allow-Origin:*" so that should be correct.
I can see from the network activity that the receiver does not try to load the vtt file. Here is the activity:
![network](https://cloud.githubusercontent.com/assets/2315801/17571879/48280438-5f52-11e6-9109-090659631d41.PNG)
And the Console is empty, apart from the usual
```
[  0.326s] [cast.receiver.CastReceiverManager] Version: 2.0.0.0044
```

**Comment by naddaf** [comment]
Did you try the CastVideos-android (v2) app?

**Comment by mathiasi** [comment]
Are you referring to https://github.com/googlecast/CastVideos-android-v2 ? I tried to import it but got an error in Android Studio
```
Gradle 'CastVideos-android-v2' project refresh failed
  Error:Configuration with name 'debug' not found.
```
and combined with the warning on the github page to use v3 instead, I have not tried it as I assumed it was outdated. But as far as i can tell, the v3 ( https://github.com/googlecast/CastVideos-android ) does not use the Companion Library.

**Comment by naddaf** [comment]
v3 is not based on the CCL, so if you want a sample of CCL project, the v2 project is the right one. I am able to compile the v2 following the insrtructions; it seems like a gradle misconfiguration on your system.
What is the url of the video file? The one that you had in the zipped project was not showing me the CORS headers when I tried that. I suggest you try this combination in your app to see if it works or not:
Track: https://commondatastorage.googleapis.com/gtv-videos-bucket/CastVideos/tracks/DesigningForGoogleCast-en.vtt

Video: https://commondatastorage.googleapis.com/gtv-videos-bucket/CastVideos/mp4/DesigningForGoogleCast.mp4

**Comment by mathiasi** [comment]
Yea I'm sorry, I forgot that I moved the video to my own host to ensure CORS was not a problem, but did not change the URL in the gist. 
But! I now seems to have figured out what was the problem as I now have captions displaying on my screen. I have never worked with WebVTT files before so I grabbed an example that I found somewhere. I wanted to make sure that the correct headers where set (origin-access, content-type etc.) so I wrapped the content of the vtt in a php file and I think that was causing the problem. I honestly did not think that it would matter but I guess it is messing with the way that the receiver loads/checks the file. It was being ignored and a Track element was not created in the Video element on the page. I guess if it is a bug, it is related to the receiver, not the companion library. I honestly would have expected some errors being thrown by the receiver but it just skipped the track.
Thank you for assisting me in resolving this problem! 👍
